### PR TITLE
Add orthographic view controller

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

## Summary
Add orthographic view controller and updated `view_control` demo to showcase the new feature

Additionally:
* added function for setting camera projection type
* updated ProjectionMatrix function to support orthographic projection
* added Project function for projecting 3d word pos to 2d screen pos.

## Test it

Build and run the `ign-rendering/examples/view_control` demo 

```
mkdir build
cd build
cmake ..
make -j 4
./view_control ogre2
```
Press `O` to toggle between Orbit View Controller and Orthographic View Controller.

ortho_view_control

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/rendering/base/BaseCamera.hh [hunk 1]
```diff
@@ -522,6 +578,37 @@ namespace ignition
...
+      math::Vector2i screenPos;
+      math::Matrix4d m = this->ProjectionMatrix() *  this->ViewMatrix();
+      math::Vector3d pos =  m * _pt;
+      double w = m(3, 0) * _pt.X() + m(3, 1) * _pt.Y() + m(3, 2) * _pt.Z()
+          + m(3, 3);
+      pos.X() = pos.X() / w;
+      pos.Y() = pos.Y() / w;
+
+      screenPos.X() = ((pos.X() / 2.0) + 0.5) * this->ImageWidth();
+      screenPos.Y() = (1 - ((pos.Y() / 2.0) + 0.5)) * this->ImageHeight();
```
Comment (chapulina) [inline include/ignition/rendering/base/BaseCamera.hh:608]: There are new Windows warnings:

```
'=': conversion from 'T' to 'T', possible loss of data
```

File: include/ignition/rendering/Camera.hh [hunk 2]
```diff
@@ -193,6 +202,24 @@ namespace ignition
...
+      public: virtual CameraProjectionType ProjectionType() const = 0;
+
+      /// \brief Set the projection type for this camera
+      /// \param[in] _matrix Camera projection type
+      public: virtual void SetProjectionType(CameraProjectionType _type) = 0;
+
+      /// \brief Project point in 3d world space to 2d screen space
+      /// \param[in] _pt Point in 3d world space
+      /// \return Point in 2d screen space
+      public: virtual math::Vector2i Project(const math::Vector3d &_pt) = 0;
```
Comment (chapulina) [inline include/ignition/rendering/Camera.hh:221]: Can this function be `const`?

File: include/ignition/rendering/OrthoViewController.hh [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+      /// \brief Set camera pan (translational movement) around target point.
+      /// \param[in] _value Pan amount in image plane, e.g. mouse drag delta.
+      public: virtual void Pan(const math::Vector2d &_value);
+
+      /// \brief Set camera orbit (rotational movement) around target point.
+      /// \param[in] _value robit amount in image plane, e.g. mouse drag delta
+      public: virtual void Orbit(const math::Vector2d &_value);
+
+      public: virtual void Resize(const unsigned int _width,
```
Comment (chapulina) [inline include/ignition/rendering/OrthoViewController.hh:78]: Missing docs

File: include/ignition/rendering/Camera.hh [hunk 4]
```diff
@@ -193,6 +202,24 @@ namespace ignition
...
+      public: virtual void SetProjectionMatrix(const math::Matrix4d &_matrix)
+          = 0;
+
+      /// \brief Get the projection type for this camera
+      /// \return Camera projection type
+      public: virtual CameraProjectionType ProjectionType() const = 0;
+
+      /// \brief Set the projection type for this camera
+      /// \param[in] _matrix Camera projection type
+      public: virtual void SetProjectionType(CameraProjectionType _type) = 0;
```
Comment (chapulina) [inline include/ignition/rendering/Camera.hh:216]: Should we document the relationship between projection type and projection matrix, and that when a type we (may?) ignore the custom projection matrix?

File: src/OrthoViewController.cc [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+//////////////////////////////////////////////////
+void OrthoViewController::SetTarget(const math::Vector3d &_target)
+{
+  this->dataPtr->target = _target;
+}
+
+//////////////////////////////////////////////////
+math::Vector3d OrthoViewController::Target() const
```
Comment (chapulina) [inline src/OrthoViewController.cc:104]: How about returning const ref so we don't force a copy?

File: src/OrthoViewController.cc [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+//  double distance =
+//      this->dataPtr->camera->WorldPosition().Distance(this->dataPtr->target);
+
+  // pseudo distance
+//   double distance = 1000.0/this->dataPtr->scale;
+
+//  double hfov = this->dataPtr->camera->HFOV().Radian();
+//  double vfov = 2.0f * atan(tan(hfov / 2.0f) /
+//        this->dataPtr->camera->AspectRatio());
```
Comment (chapulina) [inline src/OrthoViewController.cc:210]: Reminder to remove these comments if they're not needed

File: ogre2/src/Ogre2Camera.cc [hunk 7]
```diff
@@ -289,6 +289,34 @@ math::Matrix4d Ogre2Camera::ViewMatrix() const
...
+  BaseCamera::SetProjectionMatrix(_matrix);
+  this->ogreCamera->setCustomProjectionMatrix(true,
+      Ogre2Conversions::Convert(this->projectionMatrix));
+}
+
+//////////////////////////////////////////////////
+void Ogre2Camera::SetProjectionType(CameraProjectionType _type)
+{
+  BaseCamera::SetProjectionType(_type);
+  if (this->projectionType == CPT_PERSPECTIVE)
```
Comment (nkoenig) [inline ogre2/src/Ogre2Camera.cc:304]: You could use a `switch` that has a default to perspective. Something like:
```
switch (this->projectionType)
{
  default:
  case CPT_PERSPECTIVE:
    this->ogreCamera->setProjectionType(Ogre::PT_PERSPECTIVE);
    break;
  case CPT_ORTHOGRAPHIC:
    this->ogreCamera->setProjectionType(Ogre::PT_ORTHOGRAPHIC);
    break;
}

File: src/OrthoViewController.cc [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+/// \brief OrthoViewController private data.
+class ignition::rendering::OrthoViewControllerPrivate
+{
+  /// \brief Build a custom scaled orthographic projection matrix.
+  /// \param[in] _left Left position
+  /// \param[in] _right Right position
+  /// \param[in] _bottom Bottom position
+  /// \param[in] _top Top position
+  /// \param[in] _near Near clip distance
+  /// \param[in] _far Far clip distance
```
Comment (ahcorde) [inline src/OrthoViewController.cc:34]:
```suggestion
  /// \param[in] _far Far clip distance
  /// \return ...
```

File: src/OrthoViewController.cc [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+  double newOrthoWidth = width / this->dataPtr->scale;
+  double newOrthoHeight = height / this->dataPtr->scale;
+
+  translation.Set(0.0,
+      ((screenPos.X() - width/2.0) / static_cast<double>(width))
+      * newOrthoWidth,
+      ((screenPos.Y() - height/2.0) / static_cast<double>(height))
+      * newOrthoHeight);
+
+//  this->TranslateLocal(translation);
```
Comment (ahcorde) [inline src/OrthoViewController.cc:171]:
```suggestion
  //  this->TranslateLocal(translation);
```

File: src/OrthoViewController_TEST.cc [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+  EXPECT_EQ(nullptr, viewControl.Camera());
+  EXPECT_EQ(math::Vector3d::Zero, viewControl.Target());
+
+  // test setting caream
+  viewControl.SetCamera(camera);
+  EXPECT_EQ(camera, viewControl.Camera());
+
+  viewControl.SetCamera(CameraPtr());
+  EXPECT_EQ(CameraPtr(), viewControl.Camera());
+
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:70]:
```suggestion
```

File: src/OrthoViewController_TEST.cc [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+{
+  RenderEngine *engine = rendering::engine(_renderEngine);
+  if (!engine)
+  {
+    igndbg << "Engine '" << _renderEngine
+              << "' is not supported" << std::endl;
+    return;
+  }
+  ScenePtr scene = engine->CreateScene("scene");
+  EXPECT_NE(scene, nullptr);
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:53]:
```suggestion
  EXPECT_NE(nullptr, scene);
```

File: src/OrthoViewController_TEST.cc [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+
+  CameraPtr camera =  scene->CreateCamera("camera");
+  EXPECT_NE(camera, nullptr);
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:56]:
```suggestion
  EXPECT_NE(nullptr, camera);
```

File: src/OrthoViewController_TEST.cc [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+  EXPECT_EQ(math::Pose3d::Zero, initialPose);
+
+  // test orbit in viewport x (yaw)
+  viewControl.Orbit(math::Vector2d(100, 0));
+  EXPECT_GT(camera->WorldPose().Pos().X(), initialPose.Pos().X());
+  EXPECT_GT(camera->WorldPose().Pos().Y(), initialPose.Pos().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().Y(), camera->WorldPose().Rot().Y());
+  EXPECT_LT(camera->WorldPose().Rot().Z(), initialPose.Rot().Z());
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:188]:
```suggestion
  EXPECT_GT(initialPose.Pos().X(), camera->WorldPose().Pos().X());
  EXPECT_GT(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
  EXPECT_DOUBLE_EQ(initialPose.Rot().Y(), camera->WorldPose().Rot().Y());
  EXPECT_LT(initialPose.Rot().Z(), camera->WorldPose().Rot().Z());
```

File: src/OrthoViewController_TEST.cc [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+  viewControl.Pan(math::Vector2d(0, 7));
+  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
+  EXPECT_GT(camera->WorldPose().Pos().Z(), initialPose.Pos().Z());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
+  viewControl.Pan(math::Vector2d(0, -25));
+  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
+  EXPECT_LT(camera->WorldPose().Pos().Z(), initialPose.Pos().Z());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:171]:
```suggestion
  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
  EXPECT_GT(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
  viewControl.Pan(math::Vector2d(0, -25));
  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
  EXPECT_LT(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
```

File: src/OrthoViewController_TEST.cc [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+  EXPECT_EQ(math::Pose3d::Zero, initialPose);
+
+  viewControl.Orbit(math::Vector2d(-80, 0));
+  EXPECT_GT(camera->WorldPose().Pos().X(), initialPose.Pos().X());
+  EXPECT_LT(camera->WorldPose().Pos().Y(), initialPose.Pos().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().Y(), camera->WorldPose().Rot().Y());
+  EXPECT_GT(camera->WorldPose().Rot().Z(), initialPose.Rot().Z());
+
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:200]:
```suggestion
  EXPECT_GT(initialPose.Pos().X(), camera->WorldPose().Pos().X());
  EXPECT_LT(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
  EXPECT_DOUBLE_EQ(initialPose.Rot().Y(), camera->WorldPose().Rot().Y());
  EXPECT_GT(initialPose.Rot().Z(), camera->WorldPose().Rot().Z());
```

File: src/OrthoViewController_TEST.cc [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+  EXPECT_EQ(math::Pose3d::Zero, initialPose);
+
+  // test orbit in viewport y (pitch)
+  viewControl.Orbit(math::Vector2d(0, 80));
+  EXPECT_GT(camera->WorldPose().Pos().X(), initialPose.Pos().X());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
+  EXPECT_GT(camera->WorldPose().Pos().Z(), initialPose.Pos().Z());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
+  EXPECT_GT(camera->WorldPose().Rot().Y(), initialPose.Rot().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().Z(), camera->WorldPose().Rot().Z());
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:212]: initialPose first

File: src/OrthoViewController_TEST.cc [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+  camera->SetWorldPose(initialPose);
+  EXPECT_EQ(math::Pose3d::Zero, initialPose);
+
+  viewControl.Orbit(math::Vector2d(0, -90));
+  EXPECT_GT(camera->WorldPose().Pos().X(), initialPose.Pos().X());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Y(), camera->WorldPose().Pos().Y());
+  EXPECT_LT(camera->WorldPose().Pos().Z(), initialPose.Pos().Z());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().X(), camera->WorldPose().Rot().X());
+  EXPECT_LT(camera->WorldPose().Rot().Y(), initialPose.Rot().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Rot().Z(), camera->WorldPose().Rot().Z());
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:223]: initialPose first

File: src/OrthoViewController_TEST.cc [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+  EXPECT_EQ(initialPose.Pos(), camera->WorldPose().Pos());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
+  math::Matrix4d zoomProjectionMatrix = camera->ProjectionMatrix();
+  EXPECT_NE(initialProjectionMatrix, zoomProjectionMatrix);
+  viewControl.Zoom(-5);
+  math::Matrix4d zoomProjectionMatrix2 = camera->ProjectionMatrix();
+  EXPECT_EQ(initialPose.Pos(), camera->WorldPose().Pos());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
+  EXPECT_NE(initialProjectionMatrix, zoomProjectionMatrix2);
+  EXPECT_NE(zoomProjectionMatrix, zoomProjectionMatrix2);
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:135]: initialPose first

File: src/OrthoViewController_TEST.cc [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+  viewControl.Pan(math::Vector2d(2, 0));
+  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
+  EXPECT_GT(camera->WorldPose().Pos().Y(), initialPose.Pos().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
+  viewControl.Pan(math::Vector2d(-8, 0));
+  EXPECT_DOUBLE_EQ(initialPose.Pos().X(), camera->WorldPose().Pos().X());
+  EXPECT_LT(camera->WorldPose().Pos().Y(), initialPose.Pos().Y());
+  EXPECT_DOUBLE_EQ(initialPose.Pos().Z(), camera->WorldPose().Pos().Z());
+  EXPECT_EQ(initialPose.Rot(), camera->WorldPose().Rot());
```
Comment (ahcorde) [inline src/OrthoViewController_TEST.cc:155]: initialPose first

**Review by ahcorde — CHANGES_REQUESTED** [review]
in the tests try to use the same order in the EXPECT_ functions

File: src/OrthoViewController_TEST.cc (hunk 19 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 18 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc:134]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 17 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 16 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 15 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 12 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 11 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc]: 4d7d3f7

File: src/OrthoViewController_TEST.cc (hunk 10 above)
Comment (iche033) [reply src/OrthoViewController_TEST.cc:70]: 4d7d3f7

File: src/OrthoViewController.cc (hunk 9 above)
Comment (iche033) [reply src/OrthoViewController.cc]: removed line. 4d7d3f7

File: src/OrthoViewController.cc (hunk 8 above)
Comment (iche033) [reply src/OrthoViewController.cc:34]: 4d7d3f7

File: ogre2/src/Ogre2Camera.cc (hunk 7 above)
Comment (iche033) [reply ogre2/src/Ogre2Camera.cc]: done. 4d7d3f7

File: src/OrthoViewController.cc (hunk 6 above)
Comment (iche033) [reply src/OrthoViewController.cc]: removed 6e2e25f

File: src/OrthoViewController.cc (hunk 5 above)
Comment (iche033) [reply src/OrthoViewController.cc]: done. 6e2e25f

File: include/ignition/rendering/Camera.hh (hunk 4 above)
Comment (iche033) [reply include/ignition/rendering/Camera.hh:223]: added more doc on the behavior of calling this function. 6e2e25f

File: include/ignition/rendering/OrthoViewController.hh (hunk 3 above)
Comment (iche033) [reply include/ignition/rendering/OrthoViewController.hh]: added 6e2e25f

File: include/ignition/rendering/Camera.hh (hunk 2 above)
Comment (iche033) [reply include/ignition/rendering/Camera.hh]: done. 6e2e25f

File: include/ignition/rendering/base/BaseCamera.hh (hunk 1 above)
Comment (iche033) [reply include/ignition/rendering/base/BaseCamera.hh]: testing fix in 6e2e25f

Comment (iche033) [reply include/ignition/rendering/base/BaseCamera.hh]: windows build is now green as of 7a4ffa7

**Review by ahcorde — APPROVED** [review]

**Review by mjcarroll — APPROVED** [review]

**Review by nkoenig — APPROVED** [review]

**Pull request merged**